### ISSUE TITLE
How can I log 'loss' to Tensorboard every epoch instead of every iteration?

### ISSUE THREAD

**Issue by fbrad** [allenai/allennlp#2529]
**System:**
 - OS: Linux
 - Python version: 3.6.8
 - AllenNLP version: v0.8.2
 - PyTorch version: 1.0.1

**Question**
* I have recently updated to AllenNLP v0.8.2 and I've noticed that 'loss' metric is now logged every iteration to Tensorboard. How can I log 'loss' (or other metrics) to Tensorboard every epoch instead of every iteration?

![loss](https://user-images.githubusercontent.com/1182262/53097119-5f480f00-3529-11e9-953d-b6f15ce5fd0f.png)

**Comment by matt-gardner** [comment]
Hmm, didn't these used to be separated out into an "epoch metrics" section, or something?  Maybe in the refactor something got lost.  I just checked the code, and it doesn't look like there's a separate section for the metrics, like there used to be, but maybe I missed something.  @joelgrus?

**Comment by joelgrus** [comment]
sorry, can you help me understand what I'm looking at?

I'm guessing it's the case that one of these runs (which one?) was generated with old code and one with recent code?

the "bow" ones do look like they're being logged only once an epoch (those graphs look piecewise linear to me), but with a different x-axis scale than the "nobow" ones.

so I don't think the loss is being logged every iteration (at least, not based on that chart), but it's possible that the x-value is messed up. which metrics are from old code and which are from new code?

**Comment by fbrad** [comment]
@joelgrus @matt-gardner Sorry, I should have detailed the graph.

The conala*nobow model was trained with allennlp 0.8.1, and the 'loss' values were displayed one per epoch (15 epochs)
The conala*bow model was trained with allennlp 0.8.2, and the 'loss' value now seems to be logged every iteration. 

When trying to display 'loss' values from both models, Tensorboard 'squeezes' the per-epoch 'loss' values, hence the vertical green/gray line at x = 0.

**Comment by matt-gardner** [comment]
<img width="487" alt="screen shot 2019-02-20 at 7 45 21 am" src="https://user-images.githubusercontent.com/3291951/53104030-7fed7600-34e3-11e9-844f-4e4e20d3ca05.png">

Here there is an "epoch_metrics" section and a "loss" section.  The "loss" section has always been per iteration.  Looking at the code, I _think_ what happened is that "epoch_metrics" has gone away, and now what was "epoch_metrics/loss" is combined into "loss", for problematic results.  Is that right @fbrad?  Or is there still an "epoch_metrics" section?

**Comment by fbrad** [comment]
@matt-gardner 

The "epoch metrics" and "loss" sections are still being populated in the newer version (model trained with allennlp 0.8.2)
![epoch_metrics](https://user-images.githubusercontent.com/1182262/53104694-865a0d00-3538-11e9-8af2-2f12851d1e30.png)

**Comment by matt-gardner** [comment]
Is "epoch_metrics/loss" what you're looking for?  "loss" has always been per iteration.

**Comment by fbrad** [comment]
I'm confused, because looking at older models (allennlp <= 0.8.1), "eposh_metrics/loss" is indeed logged per iteration, but "loss" is displayed for each epoch only.
![loss_epoch_vs_iter](https://user-images.githubusercontent.com/1182262/53106158-26b13100-353b-11e9-9482-4fbc39dc7d7d.png)

**Comment by joelgrus** [comment]
looking at the refactor PR nothing is obviously wrong, I'll try to dig into this further today

**Comment by matt-gardner** [comment]
Oh!  Maybe "loss" used to just be validation loss, which is only computed once per epoch.  Now it looks like it's getting combined with "loss_train".  Is that it?

**Comment by matt-gardner** [comment]
Ok, yeah, I think I understand what's going on.  Something definitely changed here, thanks for pointing this out @fbrad.  @joelgrus, I'm happy to chat over slack if it's helpful.
